### ISSUE TITLE
Metadaten (Cloud Coverage etc)

### ISSUE THREAD

**Issue by nk027** [p0017/Mapping-Mining-Areas#1]
### Status Quo

- [Hier](https://github.com/p0017/Mapping-Mining-Areas/blob/d88130e560b618812b1a20c3fd0a33bdb3430b3e/gpkg_dataset_generation.py#L154) bekommen wir ein Mosaic – explicit aus dem String, implizit mit einer ID aus der Response, z.B.: `c7d9af9d-56e5-445a-b9e2-42570826594e` aus dem Field `'mosaics' > 'id'`.
- Mit einer Bounding Box bauen wir dann [hier](https://github.com/p0017/Mapping-Mining-Areas/blob/d88130e560b618812b1a20c3fd0a33bdb3430b3e/gpkg_dataset_generation.py#L179) eine Request für Quads. Eine Response ist anbei, das wichtige ist:
    – der Downloadlink, z.B.: `https://link.planet.com/basemaps/v1/mosaics/c7d9af9d-56e5-445a-b9e2-42570826594e/quads/1007-1072/full?api_key=$API_KEY`
    - und die ID aus dem Field `'items' > 'id'`, z.B.: `1007-1072`
    - (Außerdem taucht in den URLs nocheinmal die Mosaic ID auf.)

### Erweiterungen

Aus der Request für die Quads kann man under `'items' > 'links' > 'items'` die Scenes, also konkrete Bilder für die Quads ziehen. Im Browser wäre das z.B.: `https://api.planet.com/basemaps/v1/mosaics/c7d9af9d-56e5-445a-b9e2-42570826594e/quads/1006-1072/items?api_key=$API_KEY`

Hier das erste Element der Response ([response_quads.json](https://github.com/user-attachments/files/18229491/response_quads.json)) mit ID `20230609_102008_88_248f`:
```json
{
  "items": [
    {
      "link": "https://api.planet.com/data/v1/item-types/PSScene/items/20230609_102008_88_248f#ortho_analytic_4b"
    }, ...]
}
```

Jetzt brauchen wir eine andere API um Metadaten für diese Scenes (auf die wir separate keinen Zugriff haben) zu bekommen. Mit `curl` sieht das aus wie folgt:

```bash
curl -H 'Content-Type: application/json' -X POST -d '{"filter":{"type":"AndFilter","config":[{"type":"AndFilter","config":[{"type":"StringInFilter","field_name":"id","config":["20230609_102008_88_248f"]},{"type":"StringInFilter","field_name":"item_type","config":["PSScene"]}]}]},"item_types":["PSScene"]}' https://api.planet.com/data/v1/quick-search -u $API_KEY
```

In der Response ([response_scene.json](https://github.com/user-attachments/files/18229493/response_scene.json)) sind under dem Field `'features' > 'properties'` Metadaten. Primär wären `cloud_cover`und `heavy_haze_percent` interessant; das ganze Set ist:
```
    'acquired': '2023-06-09T10:20:08.888033Z',
    'anomalous_pixels': 0,
    'clear_confidence_percent': 90,
    'clear_percent': 100,
    'cloud_cover': 0,
    'cloud_percent': 0,
    'ground_control': True,
    'gsd': 3.7,
    'heavy_haze_percent': 0,
    'instrument': 'PSB.SD',
    'item_type': 'PSScene',
    'light_haze_percent': 0,
    'pixel_resolution': 3,
    'provider': 'planetscope',
    'published': '2023-06-09T13:16:58Z',
    'publishing_stage': 'finalized',
    'quality_category': 'standard',
    'satellite_azimuth': 97.9,
    'satellite_id': '248f',
    'shadow_percent': 0,
    'snow_ice_percent': 0,
    'strip_id': '6566401',
    'sun_azimuth': 57.7,
    'sun_elevation': 60.2,
    'updated': '2023-06-10T04:18:15Z',
    'view_angle': 5.1,
    'visible_confidence_percent': 68,
    'visible_percent': 100
```

In Python sollte das zirka wie folgt klappen:

```python
import requests

def planet_search(scene_id, api_key):
    url = "https://api.planet.com/data/v1/quick-search"
    payload = {
        "filter": {
            "type": "AndFilter",
            "config": [{
                "type": "AndFilter",
                "config": [{
                    "type": "StringInFilter",
                    "field_name": "id",
                    "config": [scene_id]
                }, {
                    "type": "StringInFilter",
                    "field_name": "item_type",
                    "config": ["PSScene"]
                }]
            }]
        },
        "item_types": ["PSScene"]
    }
    headers = {"Content-Type": "application/json"}
    response = requests.post(url, json=payload, auth=(api_key, ""))
    return response.json()

scene_id = "20230609_102008_88_248f"
api_key = ...
result = planet_search(scene_id, api_key)
```

### Plan

Für die zu segmentierenden Bounding Boxes müsste man die Dataset Generation also wie folgt anpassen:

- Über Bounding Boxes loopen.
    - Über in Frage kommende Basemaps loopen
        - Quads und ihre Scenes requesten
        - Metadaten der Quads (i.e. Scenes) requesten
    - Bild mit der besten Qualität (Clouds und Haze) runterladen

Außerdem sollten wir die Metadaten abspeichern um den Impact von Cloud Coverage besser quantifizieren zu können. *Für die halbjährlichen Daten sollte das genauso klappen, nur bekommen wir in dem Fall nur Metadaten (da es keine Alternativen gibt).*

**Comment by nk027** [comment]
Hab das jetzt in `scripts/scrape_metadata.R` implementiert und für Minen in Surinam getestet:

![image](https://github.com/user-attachments/assets/694fdec3-f569-4e00-9b15-427fa557f824)

To-do:
1. Error-handling könnte noch besser sein
    - IDs laufen eigtl durch, aber das Format ist nicht immer dasselbe und errors gehören caught
    - Einige wenige scenes hatten `502 Bad Gateway` und einen error, die gehören zumindest dokumentiert
3. Mit allen polygons aus der Union durchlaufen lassen und
    - die besten Quads requesten
    - Summaries und Plots über die Qualität erstellen

**Comment by p0017** [comment]
Kann ich alles nachvollziehen, aber woher nimmst du die Scene ID `20230609_102008_88_248f` ? Diese findet sich ja nirgends in `response_quads.json`. Hier findet sich ja nur die Mosaic ID `c7d9af9d-56e5-445a-b9e2-42570826594e` und die Tile ID `1007-1072`.

**Comment by p0017** [comment]
Ups, verdrückt

**Comment by nk027** [comment]
Hier unter `/items`: https://github.com/p0017/Mapping-Mining-Areas/blob/c9fcc470d16d697ecc9de583c20f5f89bb11882a/scripts/scrape_metadata.R#L94

Kannst du mir die unions senden, dann lass ich es damit laufen. Und let me know wie du den Output brauchst – Link für das richtige Quad pro Jahr o.ä. und Parquet / CSV ...

**Comment by p0017** [comment]
Habs geschafft deine Ergebnisse zu reproduzieren. Aber was mich irritiert ist, dass ich pro `basemap_id` und `tile_id` Kombi für jeden Monat mehrere Scenes zurück bekomme. Diese sind zwar alle aus dem jeweilgen Monat, aber an unterschiedlichen Tagen aufgenommen, und somit alle mit unterschiedliche Metadaten. 

Sind unsere monatlichen Tiles also Composites die aus diesen Scenes zusammengesetzt wurden? Soll ich den Monat mit der geringsten durchschnittlichen Cloud Coverage wählen? Und verstehst du den Unterschied zwischen `cloud_percent `und `cloud_cover`? Abgesehen vom Wertebereich.

![grafik](https://github.com/user-attachments/assets/af5d3601-aa4f-4919-abf3-c9609a491b07)

**Comment by nk027** [comment]
Ja, die Quads sind composites. Leider ist nicht klar welcher Part der Scenes verwendet wird, aber wir können den Quad mit den besten Scenes verwenden. (Und wir haben ein upper bound für Quality issues.) Wenn du das Shapefile schickst lass ich es durchlaufen.

- Doku ist hier <https://developers.planet.com/docs/data/skysatscene/>
- IIRC haben nicht alle `cloud_percent`, aber wenn das nur prä-monthly ist dann würde ich das + heavy haze als Entscheidungskriterium hernehmen.

**Comment by p0017** [comment]
Das wär jetzt kein großes Problem mehr, es ins `segmentation_dataset_generation.py` Skript einzufügen, aber mich besorgt die Laufzeit. 

Pro Polygon ist die Laufzeit für Scene ID Request und darauffolgenden Metadaten Request der korrespondierenden Scenes etwa 30 Sekunden. Das würde für unsere 20.000 Polygone ziemlich genau 7 Tage dauern.

Wüsste auf die Schnelle nicht, wie ich das unkompliziert parallelisieren könnte, nachdem die API Requests der Bottleneck sind. Hab auch probiert einen Scene Request über die Orders API, statt über den Data API Quick Search zu machen, um daraus vlt einen Batch Request zu basteln, bekomme aber nur Fehlermeldungen `no access to assets: PSScene/20210725_055147_22_1066/[ortho_visual] `. Ich geb dir den Orders API Request anbei. 

Hast du Ideen wie wir die Runtime drastisch verkürzen könnten? Müssten von über 30 auf etwa 4 Sekunden pro Polygon kommen, dann würde sich die Runtime zumindest nur um einen Tag verlängern.
![grafik](https://github.com/user-attachments/assets/3b39af26-b46f-42d8-a829-bc24f6e1a760)

**Comment by nk027** [comment]
Keine Ahnung, aber es muss eh nur einmal laufen und ist eigtl embarassingly paralllel, aber

- warum baust du dasselbe in Python
- > Wenn du das Shapefile schickst lass ich es durchlaufen.

**Comment by p0017** [comment]
Fixed in most recent commit.
